vicodyn/src/vicodyn/darkmetrics.cpp

```diff
@@ -0,0 +1,66 @@
+#include "cocaine/vicodyn/darkmetrics.hpp"
+
+#include <cocaine/context.hpp>
+#include <cocaine/dynamic.hpp>
+#include <cocaine/format/dynamic.hpp>
+#include <cocaine/format/exception.hpp>
+#include <cocaine/logging.hpp>
+#include <cocaine/unicorn/value.hpp>
+
+#include <blackhole/logger.hpp>
+
+namespace cocaine {
+namespace vicodyn {
+
+darkmetrics_t::darkmetrics_t(context_t& context, peers_t& peers, const dynamic_t& args)
+    : unicorn_(api::unicorn(context, args.as_object().at("unicorn", "core").as_string()))
+    , prefix_(args.as_object().at("unicorn_prefix", "/darkvoice").as_string())
+    , logger_(context.log("vicodyn/darkmetrics"))
+    , peers_(peers) {
+}
+
+darkmetrics_t::~darkmetrics_t() {
+    for (auto& scope : scopes_) {
+        scope->close();
+    }
+}
+
+auto darkmetrics_t::subscribe(const std::string& cluster) -> void {
+    if (clusters_.emplace(cluster).second) {
+        namespace ph = std::placeholders;
+        auto path = format("{}/{}/metrics", prefix_, cluster);
+        scopes_.emplace_back(unicorn_->subscribe(std::bind(&darkmetrics_t::on_subscribe, this, path, ph::_1), path));
+        COCAINE_LOG_INFO(logger_, "subscribed to metrics by path \"{}\"", path);
+    }
+}
+
+auto darkmetrics_t::on_subscribe(const std::string& path, std::future<unicorn::versioned_value_t> future) -> void {
+    /*** Message format:
+    {
+        "timestamp": 1533922332,
+        "nodes": {
+            "121170e8-a830-4239-8875-8d18307a920b": { "system_usable_weight": 0.05964477254440246 }
+        }
+    }
+    ***/
+    try {
+        auto versioned_value = future.get();
+        if (versioned_value.version() < 0) {
+            COCAINE_LOG_DEBUG(logger_, "metrics by path \"{}\" is empty", path);
+            return;
+        }
+        auto value = versioned_value.value();
+        auto timestamp = value.as_object()["timestamp"].as_uint();
+        for (const auto& peer : value.as_object()["nodes"].as_object()) {
+            auto system_usable_weight = peer.second.as_object()["system_usable_weight"].as_double();
+            peers_.set_system_usage_weight(peer.first, system_usable_weight);
+            COCAINE_LOG_DEBUG(logger_, "metrics for peer {}: system_usable_weight={} timestamp={}",
+                    peer.first, system_usable_weight, timestamp);
+        }
+    } catch (const std::exception& err) {
```

Comment:
In case of error old value of _system_usable_weight_ for peer would be used in peer selection.

Reply:
Thank you. FIxed. Old values will be changed to default after timeout.